x/distribution/spec/03_begin_block.md

```diff
@@ -0,0 +1,86 @@
+<!--
+order: 3
+-->
+
+# Begin Block
+
+At each `BeginBlock`, all fees received in the previous block are transferred to
+the distribution `ModuleAccount` account. When a delegator or validator
+withdraws their rewards, they are taken out of the `ModuleAccount`. During begin
+block, the different claims on the fees collected are updated as follows:
+
+- The block proposer of the previous round and its delegators receive between 1% and 5% of fee rewards.
+- The reserve community tax is charged.
+- The remainder is distributed proportionally by voting power to all bonded validators
+
+To incentivize validators to wait and include additional pre-commits in the block, the block proposer reward is calculated from Tendermint pre-commit messages.
```

Comment:
Looking at https://github.com/cosmos/cosmos-sdk/blob/master/x/distribution/abci.go#L31-L32

Am I reading this right that when we execute block N, the proposer for block N-1 gets the proposer reward based on the LastCommit included in block N? Which is to say the proposer of block N-1 is getting rewarded based on what the proposer of block N is doing? If this is the case there would appear to be a griefing attack where the proposer of block N just includes the minimal commit set (ie. only 67% votes rather than closer to 100%) such that the proposer of block N-1 will only get the minimum proposer reward. In practice a proposer could just not wait the TimeoutCommit before proposing, since there's technically no incentive to (ie. they'll get paid based on what the next proposer does, not what they do - the previous proposer gets paid based on what they do). Is that right or am I missing something? If it's somehow not a bug, I guess it should probably at least get some extra documentation.

On a related note, do you know of some reason all this logic needs to happen in BeginBlock of block N? Could it instead happen in the EndBlock of block N-1 ?

Reply:
Yeah, these discussions were had so long ago and not all of them were documented unfortunately, but I am 95% sure this is intentional.

See: https://github.com/cosmos/cosmos-sdk/issues/2525#issuecomment-430838701

Reply:
I think #2525 is referring to a different issue in the protocol itself, where validators can withhold votes from particular proposers to grief them. The issue I've identified seems to be an implementation bug, where proposers are getting rewarded based not on what they include in their own block, but on what the next proposer does.

Reply:
@ebuchman Why do you conclude that the sum of total votes is based on the precommits in the current block? I'm not familiar with what happens in Tendermint, but the votes are read with this line:
```
	for _, voteInfo := range req.LastCommitInfo.GetVotes() {
```
I believe that the `LastCommitInfo` contains info about the previous block. If it doesn't, I agree that it seems like a bug.

From what I gather from the issue @alexanderbez posted, that is more about how a validator can exclude other validators, eventually possibly cause them to get slashed due to liveness issues, but it would be hard to do and would come at some cost to the malicious validator. If, on the other hand, the previous proposer gets bonuses based on the current proposer's included precommits that
a) makes no sense at all, and
b) means that a malicious validator can do this "liveness" attack at 0 cost to themselves.

Can we verify that the LastCommitInfo actually passes the info from the previous block? If it doesn't I'd classify this as a bug, possibly caused by the fact that the distribution used to happen in `EndBlock` and has since been moved to `BeginBlock`.

Reply:
> I believe that the LastCommitInfo contains info about the previous block.

This is true, but those votes are included in the current block - ie. block N includes data called LastCommit, which is the commit for block N-1. The proposer for block N is responsible for putting this commit together and including it in block N, even though it's the commit for block N-1 (that's how Tendermint works and why the field is called LastCommit and not Commit). 

> possibly caused by the fact that the distribution used to happen in EndBlock and has since been moved to BeginBlock.

This seems super important and probably the source of the bug. If this was happening in EndBlock all would be well again :D

Reply:
So let's see if I have this straight.

Proposer for N-1 puts together a block: they run `BeginBlock`, the transactions, then `EndBlock`. The package it up and send it out to the other validators. This block is signed with some number of precommits, and agreed upon, and the next round starts.

Proposer for N puts together a block. They start with `BeginBlock`. Here they look at all the precommits from the previous round, package them up and include them (how?). The proposer for N-1 gets rewarded based on how many of these precommits are included.

Is that a correct understanding of what you are describing?

It seems the correct behavior is: in BeginBlock, the proposer of *the current block* gets rewarded based on how many precommits of the previous block they included.

> It would still be good to confirm this with a test, for instance have the first proposer make a block and include a commit for 100% of the voting power and then the second proposer makes a block with a commit that contains only 67% of the voting power and see how much the first proposer gets rewarded. It should be the maximum but with the current bug I believe it would end up being the minimum!

Good idea! That's beyond my expertise (and current time available) but who would be able to run a test like that?

> This seems super important and probably the source of the bug. If this was happening in EndBlock all would be well again :D

Actually, if it works the way I think (described above) it shouldn't matter, because `EndBlock` would still get run before the precommits come in, so it would be essentially the same thing. Then it's better to have it in `BeginBlock`, I think, because it doesn't depend on any of the events in the block.

Reply:
@ebuchman would you mind making this an issue? I think this PR can still be merged (really this piece of documentation seems to reflect the common understanding of what the code does or should do), and I want to make sure that we have an issue where we can keep the discussion in less precarious form.

But please reference this discussion and code location in the issue, so we remind ourselves to update it if necessary.

Reply:
Thanks opened an issue: https://github.com/cosmos/cosmos-sdk/issues/9161

> Proposer for N-1 puts together a block: they run BeginBlock, the transactions, then EndBlock. The package it up and send it out to the other validators. This block is signed with some number of precommits, and agreed upon, and the next round starts.

Note this isn't quite right. Currently a proposer puts together a block (sat N-1), but it is not executed until it is committed (ie. +2/3 precommits for it are seen). However the canonical commit is only determined by the next proposer when they put together block N and include the LastCommit.

> Proposer for N puts together a block. They start with BeginBlock. Here they look at all the precommits from the previous round, package them up and include them (how?). The proposer for N-1 gets rewarded based on how many of these precommits are included.

Again not quite, the proposer doesn't run BeginBlock before proposing. They collect precommits they've seen for block N-1 (there should be at least +2/3, but they'll wait TimeoutCommit to collect more since they're supposed to be rewarded for doing so), and they include those in the proposed block N as the LastCommit. They are supposed to be rewarded based on the size of LastCommit they include but currently the reward for that is going to the previous proposal. 

It seems correct because proposer for block N gets rewarded based on the canonical commit for block N, but the commit for block N is included in block N+1 by the proposer for N+1 so it's actually not correct to reward the proposer of N for the commit of block N, they should be rewarded for the LastCommit contained within block N (which they selected) which is the commit for block N-1. The off-by-one is super confusing :) 

> the proposer of the current block gets rewarded based on how many precommits of the previous block are included.

yes this would be correct behaviour

Reply:
Okay, still just trying to get my head around the commit process. How can the proposer for N-1 put together a block without actually executing it (BeginBlock + txs + EndBlock)? Or does the proposer for N-1 select a set of txs, bundle them, collect precommits on the bundle, then execute the block and send off the final block to everyone by executing the block? If so, the proposer for N-1 should be able to include the precommits for the current block in BeginBlock.